specs/core/0_beacon-chain.md

```diff
@@ -547,385 +534,469 @@ class BeaconState(Container):
+        # Surround vote
+        (data_1.source.epoch < data_2.source.epoch and data_2.target.epoch < data_1.target.epoch)
+    )
 ```
 
-### `get_current_epoch`
+#### `is_valid_merkle_branch`
 
 ```python
-def get_current_epoch(state: BeaconState) -> Epoch:
+def is_valid_merkle_branch(leaf: Hash, branch: Sequence[Hash], depth: int, index: int, root: Hash) -> bool:
     """
-    Return the current epoch of the given ``state``.
+    Check if ``leaf`` at ``index`` verifies against the Merkle ``root`` and ``branch``.
     """
-    return slot_to_epoch(state.slot)
+    value = leaf
+    for i in range(depth):
+        if index // (2**i) % 2:
+            value = hash(branch[i] + value)
+        else:
+            value = hash(value + branch[i])
+    return value == root
 ```
 
-### `get_epoch_start_slot`
+### Misc
+
+#### `shuffle_index`
 
 ```python
-def get_epoch_start_slot(epoch: Epoch) -> Slot:
+def shuffle_index(index: ValidatorIndex, index_count: int, seed: Hash) -> ValidatorIndex:
     """
-    Return the starting slot of the given ``epoch``.
+    Return the shuffled validator index corresponding to ``seed`` (and ``index_count``).
     """
-    return Slot(epoch * SLOTS_PER_EPOCH)
+    assert index < index_count
+    assert index_count <= 2**40
+
+    # Swap or not (https://link.springer.com/content/pdf/10.1007%2F978-3-642-32009-5_1.pdf)
+    # See the 'generalized domain' algorithm on page 3
+    for current_round in range(SHUFFLE_ROUND_COUNT):
+        pivot = bytes_to_int(hash(seed + int_to_bytes(current_round, length=1))[0:8]) % index_count
+        flip = ValidatorIndex((pivot + index_count - index) % index_count)
+        position = max(index, flip)
+        source = hash(seed + int_to_bytes(current_round, length=1) + int_to_bytes(position // 256, length=4))
+        byte = source[(position % 256) // 8]
+        bit = (byte >> (position % 8)) % 2
+        index = flip if bit else index
+
+    return ValidatorIndex(index)
 ```
 
-### `is_active_validator`
+#### `compute_committee`
 
 ```python
-def is_active_validator(validator: Validator, epoch: Epoch) -> bool:
+def compute_committee(indices: Sequence[ValidatorIndex],
+                      seed: Hash,
+                      index: int,
+                      count: int) -> Sequence[ValidatorIndex]:
     """
-    Check if ``validator`` is active.
+    Return the committee corresponding to ``indices``, ``seed``, ``index``, and committee ``count``.
     """
-    return validator.activation_epoch <= epoch < validator.exit_epoch
+    start = (len(indices) * index) // count
+    end = (len(indices) * (index + 1)) // count
+    return [indices[shuffle_index(ValidatorIndex(i), len(indices), seed)] for i in range(start, end)]
 ```
 
-### `is_slashable_validator`
+#### `validate_indexed_attestation`
 
 ```python
-def is_slashable_validator(validator: Validator, epoch: Epoch) -> bool:
+def validate_indexed_attestation(state: BeaconState, indexed_attestation: IndexedAttestation) -> None:
```

Comment:
I think `verify` is a better term to be using here. I did a bunch of searching to try back up my intuition here, best I can come up with is this from an (adopted) IEEE standard:
* Validation: The assurance that a product, service, or system meets the needs of the customer and other identified stakeholders.
* Verification. The evaluation of whether or not a product, service, or system complies with a regulation, requirement, specification, or imposed condition.

Reply:
We were distinguishing `verify` and `validate` as -- 
* `verify` was returning a bool
* `validate` has no return value and instead throws if bad input

I'm fine leaving as is..

Reply:
Fine. As long as it is consistent (and I know the rule) then I'm happy. :)